File: src/cadenza/domain/project.py
"""Project identity and the snapshot handed to a run (design doc section 3.2)."""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass

from cadenza.domain.clone_source import CloneSource

__all__ = ["FieldOrigin", "Project", "ResolvedProject"]


@dataclass(frozen=True)
class Project:
    """A project as a composed catalog holds it.

    ``project_id`` is immutable and never reused; ``aliases`` are mutable
    display names, ordered as declared and unique across the whole namespace.
    """

    project_id: str
    aliases: tuple[str, ...]
    source: CloneSource
    base_branch: str


@dataclass(frozen=True)
class FieldOrigin:
    """Which layer, and which file within it, decided one field."""

    layer: str
    file: str


@dataclass(frozen=True)
class ResolvedProject:
    """What a run persists.

    ``config_digest`` is what makes a later audit possible: a run that recorded
    only the typed name cannot tell that the catalog moved underneath it.
    """

    project_id: str
    aliases: tuple[str, ...]
    source: CloneSource
    base_branch: str
    config_digest: str
    provenance: Mapping[str, FieldOrigin]

File: src/domain/project.ts
/**
 * Project identity (design doc section 3.2).
 *
 * `projectId` is immutable and never reused; `aliases` are mutable display
 * names, ordered as declared and unique across the whole namespace.
 *
 * `ResolvedProject` -- the snapshot handed to a run, carrying `configDigest`
 * and per-field provenance -- is not here yet. It is the subject of
 * `tests/test_resolve.py`, and resolution is a later belt.
 */
import type { CloneSource } from "./clone-source.js";

export interface Project {
  readonly projectId: string;
  readonly aliases: readonly string[];
  readonly source: CloneSource;
  readonly baseBranch: string;
}

export function project(
  projectId: string,
  aliases: readonly string[],
  source: CloneSource,
  baseBranch: string,
): Project {
  return { projectId, aliases, source, baseBranch };
}
